feat(MenuItem|AccordionTitle): add ability for removing and customizing the (active|submenu) indicator

mnajdova opened the pull request:
This PR aims to fix https://github.com/stardust-ui/react/issues/687 (allowing the user to remove the submenu indicator), as well is is adding props for customizing the indicator, by providing an icon shorthand. I have few proposal for the API, and would like to hear your opinion on them.

1. Add explicit props for all things
- hideSubmenuIndicator: boolean - indicartes whether the submenu indicators should be shown
- submenuIndicatorVertical: ShorthandValue - shorthand for icon for the submenu indicators in the vertical menus
- submenuIndicatorHorizontal: ShorthandValue - shorthand for icon for the submenu indicators in the horizontal menus

Pros: this API provides you to define all indicators in the root menu, and all these will be propagated to all submenus (vertical and horizontal)
Cons: too many properties added for just one slot

2. Add single property
- submenuIndicator: boolean | ShorthandValue - this is a single prop that defined whether the submenu indicator shuold be shown (true or false), but also allows the user to specify shorthand for the icon in it. 

Props: we are adding just one prop for the submenu indicator, so the user doesn't have to think which one he should apply.
Cons: with this approach, we cannot specify on the root menu what should the submenu indicators in all menus be, as the vertical and horizontal menus are most likely to be different icons. This means, every menu shuold define the submenu indicators in it's first children MenuItems.

**3. Add single property and transform the icon** - this is currently implemented
- submenuIndicator: boolean | ShorthandValue - this is a single prop that defined whether the submenu indicator shuold be shown (true or false), but also allows the user to specify shorthand for the icon in it. For the vertical menus, we can transform this icon 90 degrees.

Props: we are adding just one prop for the submenu indicator, so the user doesn't have to think which one he should apply. Allows the user to specify the submenu indicator on the root menu
Cons: not sure whether this would be expected behavior for the client.

TODO:
- [x] update changelog
- [x] add Indicator examples

Commit:
-added option for hiding the submenu indicators as well as shorthands for the vertical and horizontal menus
-TODO rethink the names used

Commit:
-refactored menu item styles

layershifter commented on src/components/Menu/MenuItem.tsx:
```diff
@@ -193,6 +220,10 @@ class MenuItem extends AutoControlledComponent<ReactProps<MenuItemProps>, MenuIt
               defaultProps: { xSpacing: !!content ? 'after' : 'none' },
             })}
           {content}
+          {showSubmenuIndicatorIcon &&
```
Do we need `hideSubmenuIndicator` at all?

```jsx
<Menu submenuIndicatorVertical={null} submenuIndicatorVertical={null} />
```

mnajdova replied:
Nope, removed.

Commit:
-added rotate prop to the Icon
-supporting rtl for the icons
-added submenu and active indicators in the Menu and AccordionTitle components

Commit:
-removed boolean from the types of the indicators

Commit:
-added UnicodeCharacter component

Commit:
-replaced style unicode characters with UnicodeCharacter usages

Commit:
-fixed tests

mnajdova commented on src/components/Menu/MenuItem.tsx:
```diff
@@ -172,7 +178,21 @@ class MenuItem extends AutoControlledComponent<ReactProps<MenuItemProps>, MenuIt
+      vertical,
+      submenuIndicator,
+    } = this.props
+    const submenuIndicatorWithDefault = submenuIndicator === undefined ? true : submenuIndicator
```
We cannot specify the default value of the submenuIndicator in the defaultProp, as it is itemShorthand and conflicts with the children. If anybody has other idea, please let me know.

layershifter commented on src/components/Accordion/AccordionTitle.tsx:
```diff
@@ -47,26 +54,44 @@ class AccordionTitle extends UIComponent<ReactProps<AccordionTitleProps>, any> {
+  }
+
+  static defaultProps = {
+    activeIndicator: true,
```
My vote is for `indicator`, I'm not sure that we really need there a prefix

mnajdova replied:
I've added the prefix, so that people will know that the icon they are defining is regarding the active state, but yeah, we can make the name simpler.

mnajdova replied:
After some thoughts, agreed on the name indicator. We are nowhere using prefixes for the shorthands. Thanks for the suggestion!

Commit:
-replaced UnicodeCharacter with Indicator

codepretty commented on src/themes/teams/components/Icon/iconStyles.ts:
```diff
@@ -120,6 +121,8 @@ const iconStyles: ComponentSlotStylesInput<IconProps, IconVariables> = {
       ...((bordered || v.borderColor || circular) &&
         getBorderedStyles(circular, v.borderColor || getIconColor(color, v))),
+
+      transform: `${rtl ? 'scaleX(-1)' : ''} rotate(${rtl ? -1 * rotate : rotate}deg)`,
```
I love this this addition to Icon!! This is a personal preference, but I find it much easier to quickly understand what the transform is doing here if the code was written in multiple lines. eg)

```
...(rtl && {
    transform: `'scaleX(-1)' rotate(-1 * ${rotate}deg)`,
}),

...(!rtl && {
    transform: `rotate(${rotate}deg)`,
}),
```

mnajdova replied:
Agreed, will refactor it.

codepretty commented on src/themes/base/components/Indicator/indicatorStyles.ts:
```diff
@@ -0,0 +1,15 @@
+
+export default {
+  root: ({
+    props: { color },
```
Do we want an indicator to be able to specify it's own color? Since the indicator isn't a standalone component, it makes sense that it would just inherit its color from the current color. Where the current color would depend on the color set by the containing components (ie Menu, Accordion in these cases)

mnajdova replied:
Agree, will remove the color prop from this one. Thanks!

codepretty commented on src/components/Indicator/Indicator.tsx:
```diff
@@ -0,0 +1,87 @@
+        }
+      : {
+          children: Icon.create(icon, {
+            defaultProps: { color, rotate: Indicator.directionMap[direction].rotation },
```
Seems like this is making the assumption that all icons will always be rotated in the same way

mnajdova replied:
I myself had doubts on this, but I don't see any other way how we can easily achieve this (from the other components it's not good practice to change some props for the icon as it is shorthand for another component - Indicator in this case). So we can either provide prop in the Indicator for rotating (similar that we have in the Icon), or we can have assumption that the indicator icon provided will always be the bottom one (and depending on the direction we can rotate it accordingly). I am open to other proposals as well.

miroslavstastny replied:
I would prefer something like this:
```jsx
overrideProps: ({rotate}) => ({rotate: (Indicator.directionMap[direction].rotation || 0) + (rotate || 0)})
```

Then I can use `send` icon as an indicator:
```jsx
<Indicator icon={{name: "send", rotate: 90}} direction="forward" />
```

sophieH29 commented on src/components/Indicator/Indicator.tsx:
```diff
@@ -0,0 +1,87 @@
+
+export interface IndicatorProps extends UIComponentProps, ColorComponentProps {
+  /** The indicator can be direction in different directions. */
+  direction?: 'forward' | 'back' | 'top' | 'bottom'
```
For me, it seems a bit confusable, what is 'bottom' and what is 'back'. I guess it's left but seems not very intuitive.
Maybe just 'right', 'left', 'top', 'bottom' - simple and straightforward

layershifter replied:
`right` and `left` are not RTL safe.

mnajdova replied:
Would like to avoid the right and left, as in rtl it won't make sense to use them.. I am open to other suggestion. This one were gathered from the browser arrows for navigation (back and forward). Maybe start and end? What do you think?

sophieH29 replied:
Well, we use 'right' and 'left' for the Popup too.
I don't think `start`  `end` is better as it more describes a position in regards to something. If no other ideas then let's leave 'back' and 'forward'. If treat it as video player buttons it makes more sense 😄

miroslavstastny replied:
I vote for `start` and `end` as these are the direction names we use.

mnajdova replied:
@miroslavstastny agree, let's strive for consistency, although it seems really strange for me to have the '**<**' indicator a start, and '**>**' as end.. We can change it later anyway...

sophieH29 commented on src/components/Indicator/Indicator.tsx:
```diff
@@ -0,0 +1,87 @@
+        }
+      : {
+          children: Icon.create(icon, {
+            defaultProps: { color, rotate: Indicator.directionMap[direction].rotation },
```
Don't you need to get `direction` here based on rtl, same way as you did above `this.getDirectionBasedOnRtl(rtl, direction)`?

mnajdova replied:
Nope, as the Icon itself is handling rtl, if the theme wants to support that (this changes were added as part of this PR). Let me know if you think otherwise..

sophieH29 replied:
If Icon is handling it then it's totally fine

Commit:
-renamed indicator* props to indicator
-refactored icon styles to be more readable

Commit:
-removed color prop in the indicator
-added examples

Commit:
-renamed forward and back direction to start and end
-moved the rotate icon prop from default to override props in the Indicator component

Commit:
Merge branch 'master' into feat/menu-item-hide-submenu-indicator

# Conflicts:
#	src/components/Menu/MenuItem.tsx
#	src/themes/teams/components/Menu/menuItemStyles.ts

Commit:
-updated changelog

Commit:
-remove default props in AccordionTitle

Commit:
-removed arrow unicode chars from the teams theme
-remove helper methods for getting the mapping arrow in rtl

layershifter commented on docs/src/examples/components/Indicator/Types/IndicatorExampleDirection.shorthand.tsx:
```diff
@@ -0,0 +1,11 @@
+import { Indicator } from '@stardust-ui/react'
+
+const IndicatorExampleDirection = () => (
+  <div>
```
May be `Fragment` API?

```suggestion
  <>
```

layershifter commented on docs/src/examples/components/Indicator/index.tsx:
```diff
@@ -0,0 +1,10 @@
+import Types from './Types'
+
+const IndicatorExamples = () => (
+  <div>
```
```suggestion
  <>
```
We can use there Fragment API, `div` is a legacy from SUIR

layershifter commented on src/components/Icon/Icon.tsx:
```diff
@@ -37,6 +37,9 @@ export interface IconProps extends UIComponentProps, ColorComponentProps {
   name?: string
 
+  /** An icon can be rotated by the degree specified as number. */
+  rotate?: number
```
String looks like a valid value, too:

```jsx
<Indicator rotate='90' />
```

mnajdova replied:
The rotate property of the icon can be used for calculation as well, take a look in the Indicator component (the overrideProps function for the Icon). In that case we will need to cast this to number if it was provided as string. Do we really want this?

layershifter commented on src/components/Indicator/Indicator.tsx:
```diff
@@ -0,0 +1,89 @@
+          dangerouslySetInnerHTML: {
+            __html: hexUnicode && this.isHex(hexUnicode) ? `&#x${hexUnicode};` : '',
+          },
+        }
```
This looks very strange for me, I played with this an it seems that we don't need this at all: https://codesandbox.io/s/1zxj6x8ok3

I've pinned `react@16.3` and `react-dom@16.3` to ensure that it's supported and even tested in IE 😼 

---

![image](https://user-images.githubusercontent.com/14183168/51472435-2548e900-1d82-11e9-8d60-0be28d426494.png)

IE compatible build: https://create-react-app-i4qwctbvw.now.sh/

mnajdova replied:
Yep, refactored. At the start I had here string, that's why I had to do this magic, totally valid and more clear code. Thanks!

Commit:
Merge branch 'master' into feat/menu-item-hide-submenu-indicator

# Conflicts:
#	CHANGELOG.md

Commit:
-addressed comments on the PR

Commit:
-replaced div with fragment

mnajdova committed:
Merge branch 'master' into feat/menu-item-hide-submenu-indicator
